### PR TITLE
added auth

### PR DESCRIPTION
**Pull request by IemetsO**
привіт! 
я не зовсім розумію як зробити правильно форми для реєстрації та входу(
поки розбираюся як то все має працювати)

### CODE REVIEW

File: app/cards/[id]/charts/constant.ts [hunk 1]
```diff
@@ -1,145 +1,429 @@
...
-        height_lower:98,
-      },
-      {
-        age: 58,
-        height_upper:124,
-        height_lower:99,
-      },
-  ]
\ No newline at end of file
+export const HeightAndWeightForAgeMinus2 = [
```
Comment (lluchkaa) [inline app/cards/[id]/charts/constant.ts:1]: Usually I prefer `CAPITAL_SNAKE_CASE` for similar constants, but it's fine!

File: app/cards/[id]/charts/page.tsx [hunk 2]
```diff
@@ -3,7 +3,12 @@ import { Chart } from "@/components/chart"
...
 import { CardRecord } from "@/domains/cards/types"
-import { DEFAULT_HEIGHTforAGE } from "./constant"
+import {
+  HeightAndWeightForAgePlus3,
+  HeightAndWeightForAgeMinus2,
+  WeightForHeightMinus2,
+  WeightForHeightPlus2,
+} from "./constant"
 
 export default function ChartItem() {
```
Comment (anastasiia-tarasenko) [inline app/cards/[id]/charts/page.tsx]: It is not `ChartItem`, it is `ChartsPage`

File: app/layout.tsx [hunk 3]
```diff
@@ -18,13 +19,15 @@ export default function RootLayout({
...
     <html lang="en">
-      <body className={inter.className}>
-        <Header />
-        <div className="px-10">
-          {children}
+      <html lang="en">
+        <body className={inter.className}>
+          <AuthProvider>
+            <Header />
+            <div className="px-10">{children}</div>
```
Comment (anastasiia-tarasenko) [inline app/layout.tsx]: It is better to set `p-10`, so we will have spaces for all sides

File: domains/auth/services.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    const errorMessage = error.message;
+  })};
+
+
+export async function getUser (uid){
+    const userDocRef = doc(userCollection, uid)
+    const userDoc = await getDoc(userDocRef);
+   return userDoc
+   
+}
```
Comment (lluchkaa) [inline domains/auth/services.ts]: Please, check the following code, and try to use it:
```tsx
export async function login(email: string, password: string) {
  try {
    const authUser = await signInWithEmailAndPassword(auth, email, password)
    const user = await getUser(authUser.user.uid)

    return user
  } catch (error) {
    return null
  }
}

export async function getUser(uid: string) {
  const userDocRef = doc(userCollection, uid)
  const userSnapshot = await getDoc(userDocRef)

  return {
    id: userSnapshot.id,
    ...userSnapshot.data(),
  } as User
}
```

File: app/cards/[id]/charts/page.tsx
Comment (anastasiia-tarasenko) [inline app/cards/[id]/charts/page.tsx:1]: Added notes about this page to Notion.

image

File: app/login/page.tsx
Comment (anastasiia-tarasenko) [inline app/login/page.tsx:1]: Also added comments about this page to Notion.

image

File: app/login/page.tsx [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    },
+  })
+  const { getFieldProps } = formik
+  return (
+    <div className="mt-5 text-center">
+      <form className="flex-col  " onSubmit={formik.handleSubmit}>
+        <Input {...getFieldProps("name")} label="Ім'я" type="text" />
+        {formik.touched.name && formik.errors.name ? (
+          <div className="text-red-500 text-xs">{formik.errors.name}</div>
+        ) : null}
```
Comment (lluchkaa) [inline app/login/page.tsx]: Let's try to pass `error` to `Input`.

`Input.tsx`
```tsx
import { cn } from "@/utils"
import React from "react"

type Props = React.InputHTMLAttributes<HTMLInputElement> & {
  label?: React.ReactNode
  error?: React.ReactNode
}

const _Input = (props: Props, ref: React.ForwardedRef<HTMLInputElement>) => {
  const { label, className, error, ...restProps } = props

  return (
    <div className={cn("flex flex-col items-center", className)}>
      <label>{label}</label>
      <input
        {...restProps}
        ref={ref}
        className="border-2 border-sky-400 h-10 w-2/3"
      />
      {error ? <p className="text-red-500 text-xs">{error}</p> : null}
    </div>
  )
}

export const Input = React.forwardRef(_Input)
```

Also, please, pay attention to the forwarded ref. [Read more](https://react.dev/reference/react/forwardRef).

File: app/login/page.tsx [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    }),
+    onSubmit: (values) => {
+      login(values.email, values.password)
+      router.push(`/cards`)
+    },
+  })
+  const { getFieldProps } = formik
+  return (
+    <div className="mt-5 text-center">
+      <form className="flex-col  " onSubmit={formik.handleSubmit}>
```
Comment (anastasiia-tarasenko) [inline app/login/page.tsx]: You can add `className="flex flex-col gap-5"` and you will not need `mt-5` for each input anymore.

File: app/login/page.tsx [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        <Input {...getFieldProps("name")} label="Ім'я" type="text" />
+        {formik.touched.name && formik.errors.name ? (
```
Comment (anastasiia-tarasenko) [inline app/login/page.tsx]: You can move it to a separate function and add it to services, like:

```tsx
const getErrorMessage = (key: string) => touched[key] && errors[key] ? errors[key] : null
```

After that, it is better to create separate component for error message, like:

```tsx
import React from 'react'
import { cn } from '@/utilities/cn'

type Props = React.PropsWithChildren<{
  className?: string
}>

export const ErrorMessage = (props: Props) => {
  const { children, className } = props

  return children ? (
    <p className={cn('text-red-500 text-xs', className)}>
      {children}
    </p>
  ) : null
}

```

After that, you can use it like this:

```tsx
<ErrorMessage>
  {getErrorMessage('name')}
</ErrorMessage>
```

File: components/chart/index.tsx [hunk 8]
```diff
@@ -1,45 +1,44 @@
...
-  defaultDataKey1,
-  defaultDataKey2,
-}: Props) => {
+function detectionOfColorOfLine(datas: { datas: [] }, data: []) {
+  if (datas.datas.indexOf(data) === 0) {
+    return "green"
+  } else {
+    return "red"
+  }
+}
```
Comment (lluchkaa) [inline components/chart/index.tsx]: Actually you may not need this function. See the below comment.

File: components/chart/index.tsx [hunk 9]
```diff
@@ -1,45 +1,44 @@
...
-      />
+      {DataRecord.data.datas.map((data: []) => (
+        <Line
+          key={DataRecord.data.datas.indexOf(data)}
+          type="monotone"
+          data={data}
+          dataKey={DataRecord.data.xKey}
+          stroke={detectionOfColorOfLine(DataRecord.data, data)}
+        />
+      ))}
```
Comment (lluchkaa) [inline components/chart/index.tsx]: To apply color based on `index`, you can use second parameter of `.map` callback.
See the `stroke` prop:

```tsx
datas.map((data, index) => (
  <Line
    key={DataRecord.data.datas.indexOf(data)}
    type="monotone"
    data={data}
    dataKey={DataRecord.data.xKey}
    stroke={index === 0 ? 'green' : 'red'}
  />
```

File: components/chart/index.tsx [hunk 10]
```diff
@@ -1,45 +1,44 @@
...
-        data={defaultData}
-        dataKey={defaultDataKey2}
-        stroke="red"
-      />
+      {DataRecord.data.datas.map((data: []) => (
+        <Line
+          key={DataRecord.data.datas.indexOf(data)}
+          type="monotone"
+          data={data}
+          dataKey={DataRecord.data.xKey}
```
Comment (lluchkaa) [inline components/chart/index.tsx]: Please, name the first parameter as `props`, not `DataRecord`. Also, use object restructuring here:
```tsx
type Props = {
  datas: Record<string, number>[][]
  xKey: string
  yKey: string
}

export const Chart = (props: Props) => {
  const { datas, xKey, yKey } = props

  return (
    <LineChart
      width={500}
      height={300}
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
      className="text-xs"
    >
      {datas.map((data, index) => (
        <Line
          key={index}
          type="monotone"
          data={data}
          dataKey={xKey}
          stroke={index === 0 ? "green" : "red"}
        />
      ))}
```

**Review by lluchkaa — APPROVED** [review]

File: app/login/page.tsx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    validationSchema: Yup.object({
+      email: Yup.string().required("Обов'язкове поле"),
+      password: Yup.string().min(7).required("Обов'язкове поле"),
+    }),
+    onSubmit: (values) => {
+      login(values.email, values.password)
+      router.push(`/cards`)
+    },
+  })
+  const { getFieldProps } = formik
```
Comment (anastasiia-tarasenko) [inline app/login/page.tsx]: Use destruction:

```tsx
const { getFieldProps, errors, touched } = formik
```

File: app/login/page.tsx [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+            {...getFieldProps("password")}
+            label="Пароль"
+            type="password"
+          />
+          {formik.touched.password && formik.errors.password ? (
+            <div className="text-red-500 text-xs">{formik.errors.password}</div>
+          ) : null}
+        </div>
+
+        <div>
```
Comment (anastasiia-tarasenko) [inline app/login/page.tsx]: You don't need `div` here.

File: app/login/page.tsx [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+          <Button type="submit" className="mb-6 mt-5">
+            Створити особистий кабінет
+          </Button>
+        </div>
+      </form>
+      <h4>У Вас вже є особистий кабінет? </h4>
+
+      <form className="flex-col  " onSubmit={formikSignIn.handleSubmit}>
```
Comment (anastasiia-tarasenko) [inline app/login/page.tsx]: `flex-col` doesn't work without `flex`. You can do the same here - `className="flex flex-col gap-5"`

File: app/login/page.tsx [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+            {...formikSignIn.getFieldProps("password")}
+            label="Пароль"
+            type="password"
+          />
+          {formik.touched.password && formik.errors.password ? (
+            <div className="text-red-500 text-xs">{formik.errors.password}</div>
+          ) : null}
+        </div>
+
+        <div>
```
Comment (anastasiia-tarasenko) [inline app/login/page.tsx:48]: You don't need `div` here

File: app/page.tsx [hunk 15]
```diff
@@ -22,9 +22,12 @@ export default function About() {
       </div>
       <div className="flex flex-row">
         <AboutItem>
-          Особистий кабінет із можливістю вносити показники ваги, зросту дитини
-          та отримані щеплення
+          <Link href="login">
+            Особистий кабінет із можливістю вносити показники ваги, зросту
+            дитини та отримані щеплення{" "}
```
Comment (anastasiia-tarasenko) [inline app/page.tsx]: Remove `{" "}`

File: contexts/auth/provider.tsx [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+type Props = {
+  children?: React.ReactNode
+}
+
+export function AuthProvider({ children }: Props) {
+  const [contextValue, setContextValue] = useState<AuthContextType>({
+    firebaseUser: undefined,
+    user: undefined,
+  })
+  console.log(contextValue)
```
Comment (anastasiia-tarasenko) [inline contexts/auth/provider.tsx]: Remove `console.log(contextValue)`

**Review by anastasiia-tarasenko — APPROVED** [review]
Cool! Let's move all data to Firebase next week 🤩

**Pull request merged**